src/mapped-types.ts

```diff
@@ -138,3 +138,29 @@ export interface _DeepReadonlyArray<T> extends ReadonlyArray<DeepReadonly<T>> {}
 export type _DeepReadonlyObject<T> = {
   readonly [P in keyof T]: DeepReadonly<T[P]>
 };
+
+/**
+ * DeepRequired
+ * @desc Required that works for deeply nested structure
+ */
+export type DeepRequired<T> = T extends any[]
+  ? _DeepRequiredArray<T[number]>
+  : T extends object ? _DeepRequiredObject<T> : T;
+
+/**
+ * DeepRequiredArray
+ * @desc Nested array condition handler
+ * @private
+ */
+// tslint:disable-next-line:class-name
+export interface _DeepRequiredArray<T>
+  extends Array<DeepRequired<NonNullable<T>>> {}
```

Comment:
I'm not sure about `NonNullable`. For me required is to only make prop to be required to exist and removing null from the type is not really necessary.
Imagine you have something like e.g. `{ required: string | null, optional?: string | undefined }`, I would expect the result `{ required: string | null, optional: string }` but here it would undesirably purge the null.
I would rather go with `NonUndefined` instead for a start, because null is still a defined value and IMO because it's working on deep structures there will be a lot of cases where preserving null on some props would be desired.

`type NonUndefined<T> = T extends undefined ? never : T;`

Thoughts?

Reply:
You are right. I think maybe `DeepNonNullable` is more suitable for handling both `null` and `undefined`. I could create a new PR for it later.